### DISCUSSION TITLE
Lake of tutorials

### DISCUSSION THREAD

**Discussion by hghalebi** [sknetwork-team/scikit-network#551] [General]
I am looking for toturial. It is really difficult find any real toturial and official doc are not really practical. Is there in source for this? 
@tbonald

**Comment by TiphaineV** [comment]
Hello @hghalebi, there are some tutorials as notebooks on the documentation website (section "Tutorials") : https://scikit-network.readthedocs.io/en/latest/tutorials/gnn/gnn_classifier.html; to some extents, the "Use cases" section also serves as tutorials.

@tbonald and @simondelarue also regularly organize sessions that walk through the above notebooks (the last one was on Dec. 13).

**Comment by tbonald** [comment]
Thanks for the reply, @TiphaineV!
 @hghalebi, if there is anything you would like to see in the [tutorials](https://scikit-network.readthedocs.io/en/latest/tutorials/data/index.html) that is not available, please tell us, we'll do our best to update them.

**Comment by hghalebi** [comment]
# what I was looking for
I wanted to use the graph to analyze the sensors logs of trains. I needed to iterate on my tabular data to and by my customize function, add nodes, and link to my graph. So I was looking for a method that enabled me to add edge with more control. 

And this is my google search for the tutorial that I was looking for

## Google search 
### 1. This is when I google in a general way;
![image](https://user-images.githubusercontent.com/38727229/212910938-67545369-7e38-4b6c-9298-60425ff12ec6.png)

### 2- Then I tried to be more specified: 
![image](https://user-images.githubusercontent.com/38727229/212911537-ee741636-5cb6-4a5b-8533-8062071a2de2.png)

The result is general documentation on visualization, different from what I looked for, and then documentation from another library. 

### 3- Again, I tried to be more specific on the library: 
![image](https://user-images.githubusercontent.com/38727229/212912043-19c73d45-df22-493f-a6bd-e981410190ee.png)

## What I found:
The More related document that I found was this https://scikit-network.readthedocs.io/en/latest/tutorials/data/load_data.html

# What I would love to see in the documentation

## A better SEO

Even if there was enough documentation, if those would not indexed by search engines, we could not find them. Documents that are not indexed in google ( or other search engines) are like they don't exist.

## A conceptual manifesto

I tried to understand the logic behind the library's abstraction, but it needs to be more explicit. A manifesto/whitepaper that explains the logic of abstraction and the conceptual framework of this library will be really helpful.

## Multy level of documentation.

The actual documentation helped me to understand that we could do everything possible with another framework with this framework. However, if I could have complete use cases both in a simple but low-level and more complex more significant project more in high level. 


**Also, thank you for your reactiveness and open ness to feedback**

**Comment by tbonald** [comment]
OK, thanks for your feedback.

As for your problem (adding nodes / links), there is indeed no such functions in scikit-network. The reason is that graphs are represented as sparse CSR matrices and that adding nodes / links is time-consuming. The best is to come back to an edge list, as shown here:

```
from sknetwork.data import karate_club, from_edge_list

adjacency = karate_club() # sparse CSR matrix
edge_list = [(i, j) for i, j in zip(adjacency.tocoo().row, adjacency.tocoo().col)] # edge list
edge_list += [(4, 10), (10, 34)]  # add new nodes / links
adjacency = from_edge_list(edge_list) # new sparse CSR matrix
```

We will add the corresponding function, hoping that it is later referenced by search engines. 

Thanks for the other suggestions (in particular, the manifesto and the multi-level documentation).
